src/main/java/server/entities/Facility.java

```diff
@@ -19,78 +18,114 @@ public Facility(String name){
+
     }
 
     public void removeBooking(Booking booking){
-        LocalDate bookingDate =booking.getStart().toLocalDate();
-        ArrayList bookingsList = this.bookingsTable.get(bookingDate);
+        String bookingDay = booking.getDay();
+        ArrayList bookingsList = this.bookingsTable.get(bookingDay);
         for (int i =0; i< bookingsList.size(); i++){
             Booking b = (Booking) bookingsList.get(i);
             if (b.getBookingId() == booking.getBookingId()){
                 bookingsList.remove(b);
-                return;
+//                return null;
             }
         }
-        System.out.println("Booking to remove is not found");
+//        Exception e = new NoSuchElementException("Unable to find booking to remove");
+//        return e;
     }
 
-
-
-    public ArrayList<FacilityObserver> getObserverList() {
-        return observerList;
+    public void offsetBooking(Booking booking, LocalTime newStart, LocalTime newEnd){
+        String bookingDay = booking.getDay();
+        ArrayList bookingsList = this.bookingsTable.get(bookingDay);
+        for (int i =0; i< bookingsList.size(); i++){
+            Booking b = (Booking) bookingsList.get(i);
+            if (b.getBookingId() == booking.getBookingId()){
+                b.setStart(newStart);
+                b.setEnd(newEnd);
+            }
+        }
     }
 
-    public void attach(FacilityObserver o){
-        //TODO
+    public void extendBooking(Booking booking, LocalTime newEnd){
+        String bookingDay = booking.getDay();
+        ArrayList bookingsList = this.bookingsTable.get(bookingDay);
+        for (int i =0; i< bookingsList.size(); i++){
+            Booking b = (Booking) bookingsList.get(i);
+            if (b.getBookingId() == booking.getBookingId()){
+                b.setEnd(newEnd);
+            }
+        }
     }
 
-    public void notifyObservers(){
-        //TODO
+    public void addObserver(FacilityObserver o){
+        observerList.add(o);
     }
 
-    public ArrayList<Booking> getBookings (LocalDate date){
-        return this.bookingsTable.get(date);
-    }
+    public ArrayList<FacilityObserver> getUpdatedObservers(){
 
+        ArrayList<FacilityObserver> observersToRemove = new ArrayList<>();
 
-    public void setObserverList(ArrayList<FacilityObserver> observerList) {
-        this.observerList = observerList;
+        for (FacilityObserver o: observerList){
+            if (!isValidObserver(o)){
+                observersToRemove.add(o);
+            }
+        }
+
+        for (FacilityObserver o: observersToRemove){
+            observerList.remove(o);
+        }
+
+        return observerList;
     }
 
-    public String getName() {
-        return name;
+    private Boolean isValidObserver(FacilityObserver o){
+
+        if (o.getEndDate().compareTo(LocalDateTime.now()) <0 ){
+            return false;
+        }
+        else{
+            return true;
+        }
+
     }
```

Comment:
not an issue, just a tip for code cleanliness: 

`return (o.getEndDate().compareTo(LocalDateTime.now()) >= 0 )`